### ISSUE TITLE
browsergrad-jit rank-1 matmul backward drops operand gradients

### ISSUE THREAD

**Issue by vijayksingh** [unlocalhosted/browsergrad#10]
## Summary

Published `@unlocalhosted/browsergrad-jit` 0.8.2 accepts rank-1 input in `torch.nn.functional.linear` forward, but backward leaves both the input and weight gradients as `None`. Bias gradient is populated. Rank-2 and rank-3 inputs produce all expected gradients.

This is dangerous for PyTorch-shaped teaching code because unbatched `nn.Linear` / `F.linear` appears to train while its load-bearing operands receive no gradient.

## Reproduction

```python
import torch

x = torch.nn.Parameter(torch.tensor([1.0, 2.0], dtype=torch.float32))
weight = torch.nn.Parameter(torch.tensor(
    [[1.0, 0.0], [0.0, 1.0], [1.0, 1.0]],
    dtype=torch.float32,
))
bias = torch.nn.Parameter(torch.tensor([0.5, -0.5, 1.0], dtype=torch.float32))

out = torch.nn.functional.linear(x, weight, bias)
out.sum().backward()

print(out.shape)   # (3,)
print(x.grad)      # None; expected [2.0, 2.0]
print(weight.grad) # None; expected three rows of [1.0, 2.0]
print(bias.grad)   # [1.0, 1.0, 1.0]
```

With `x` shaped `(N, 2)` or `(B, N, 2)`, input, weight, and bias gradients are populated correctly.

## Expected behavior

Match PyTorch's matmul/linear contract for rank-1 input, including vector promotion during backward and reduction back to the original operand shapes. If rank-1 VJP is intentionally unsupported, forward or backward should fail explicitly rather than silently returning missing operand gradients.

## Source observation

Current 0.9.0 source still has a MATMUL VJP guard requiring at least 2-D inputs, so the published behavior is not merely an old wheel mismatch.

## Learner impact

A custom Linear Unit needs to teach both unbatched vector inference and gradient correctness. Until fixed, the course can execute rank-1 forward parity but must restrict gradient authority to rank-2+ inputs and disclose the runtime boundary.

**Comment by vijayksingh** [comment]
> *This was generated by AI during triage.*

## Agent Brief

**Category:** bug
**Summary:** Implement correct backward semantics for rank-one matmul and linear operands.

**Current behavior:**
The symbolic matmul VJP returns no operand gradients when either operand rank is below two. A fresh `F.linear` check with a rank-one input produces the correct forward output and bias gradient, but both input and weight gradients are `None`.

**Desired behavior:**
Matmul backward should follow PyTorch shape semantics for vector-vector, vector-matrix, matrix-vector, and batched promoted cases. Rank-one operands and upstream gradients should be treated with the necessary temporary dimensions for VJP calculation, then reduced/squeezed back to each original operand shape. `F.linear` with a one-dimensional input must inherit the corrected behavior.

**Key interfaces:**
- Matmul VJP — produce gradients for every differentiable supported rank combination.
- Broadcast reduction — reduce promoted/batched gradients to exact original operand shapes.
- `F.linear()` — preserve bias handling while receiving matmul input/weight gradients.

**Acceptance criteria:**
- [ ] Vector-vector matmul backward returns both vector gradients with exact input shapes.
- [ ] Vector-matrix and matrix-vector backward return correct gradients for both operands.
- [ ] Rank-one `F.linear` returns the expected input gradient, one `[1,2]` weight-gradient row per output, and an all-ones bias gradient for a summed output.
- [ ] Broadcasted/batched matmul gradients still reduce to original operand shapes.
- [ ] Supported rank-one cases never silently return `None`; unsupported cases fail explicitly.

**Out of scope:**
- Adding new matmul dtypes or backend capabilities.
- Changing forward broadcasting semantics unrelated to the missing gradients.

**Comment by vijayksingh** [comment]
> *This was generated by AI during triage.*

## Verified outcome

**State:** wontfix — already implemented

Implemented in https://github.com/unlocalhosted/browsergrad/commit/4d7181a9e14ba8c69e8714277da4172b361afcf4: MATMUL VJP promotes vector operands and upstream gradients, computes symbolic gradients, then unbroadcasts/reshapes to original ranks. Dot, vector-matrix, matrix-vector, batched vector-matrix, and rank-one F.linear regressions pass. Fresh verification passed JIT build, both typechecks, lint, 27 unit tests, 345 integration tests, architecture checks, packed-release tests, and the required JIT semantic-permute WebGPU gate.

Closing as completed and replacing the ready state with `wontfix` per the repository triage state machine.
